src/sinks/loki/sink.rs

```diff
@@ -163,35 +163,100 @@ pub(super) struct EventEncoder {
 
 impl EventEncoder {
     fn build_labels(&self, event: &Event) -> Vec<(String, String)> {
-        let mut static_vec: HashMap<String, String> = HashMap::new();
-        let mut dynamic_vec: HashMap<String, String> = HashMap::new();
+        let mut static_labels: HashMap<String, String> = HashMap::new();
+        let mut dynamic_labels: HashMap<String, String> = HashMap::new();
 
         for (key_template, value_template) in self.labels.iter() {
-            if let (Ok(key), Ok(value)) = (
-                key_template.render_string(event),
-                value_template.render_string(event),
-            ) {
-                if let Some(opening_prefix) = key.strip_suffix('*') {
-                    let output: Result<serde_json::map::Map<String, serde_json::Value>, _> =
-                        serde_json::from_str(value.as_str());
-
-                    if let Ok(output) = output {
-                        // key_* -> key_one, key_two, key_three
-                        // * -> one, two, three
-                        for (k, v) in output {
-                            dynamic_vec.insert(
-                                slugify_text(format!("{}{}", opening_prefix, k)),
-                                Value::from(v).to_string_lossy().into_owned(),
-                            );
-                        }
+            let key = key_template.render_string(event);
+            let value = value_template.render_string(event);
+
+            if key.is_err() || value.is_err() {
+                Self::warn_on_label_expansion_error(key_template, value_template, &key, &value);
+                continue;
+            }
+
+            let key_s = key.unwrap();
+            let value_s = value.unwrap();
+
+            if let Some(opening_prefix) = key_s.strip_suffix('*') {
+                let output: Result<
+                    serde_json::map::Map<String, serde_json::Value>,
+                    serde_json::Error,
+                > = serde_json::from_str(value_s.clone().as_str());
+
+                if output.is_err() {
+                    warn!(
+                        "Failed to expand dynamic label. value: {}, err: {}",
+                        value_s,
+                        output.err().unwrap()
+                    );
+                    continue;
+                }
+
+                // key_* -> key_one, key_two, key_three
+                // * -> one, two, three
+                for (k, v) in output.unwrap() {
+                    let key = slugify_text(format!("{}{}", opening_prefix, k));
+                    let val = Value::from(v).to_string_lossy().into_owned();
+                    if val == "<null>" {
+                        warn!("Encountered \"null\" value for dynamic label. key: {}", key);
                     }
```

Comment:
Seeing this explicitly, I wonder if we want to included values that are `null` or... if I'm misreading this and we aren't including them it could be more clear.

Reply:
Currently `null` values which can encounter during dynamic label expansion are included in the labels. These are included with the `<null>` string from the conversion as value.

If we want that is a good question, I tried to only add the warnings without changing the current behaviour. As warnings will be emitted we could as well `continue` in such cases and no longer populate these labels.
From a user perspective I think it would be fine to drop these labels when there are warnings within the logs.

Please let me know which approach you prefer.

Reply:
For now I've added the `continue` and skip the null values. In case it is preferred to keep the values included it can be removed again.

Rebased the PR branch on `master`.